keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,452 @@
+
+1) To get rid of our reliance on GOPATH hackery to build Kubernetes and to run
+   verifier tools and code-generators.
+2) To simplify and clarify our tooling with regards to multi-module operation.
+3) To be able to use upstream tooling, such as `gopls` across multiple modules.
+
+### Non-Goals
+
+1) To rewrite our code-generators (though some could use it).
+2) To make the build or tools faster.
+
+## Proposal
+
+This KEP proposes to add a `go.work` file (checked in) and fix all the tools
+that break because of that.  At the end, there will be no reliance on GOPATH.
+verify and update scripts can use normal Go conventions for building and
+running code.  Abominations like `run-in-gopath.sh` will be deleted.
+
+The implementation will come in several waves, which will have to be sequenced
+across k8s.io/kubernetes, k8s.io/gengo, and k8s.io/kube-openapi repos.  In
+order to not leave repos in a broken state, some PRs will need to be larger
+than we might otherwise like, but commits can tell the story.
+
+This is a proposed merge sequence:
+
+1. Retool the core of gengo to handle workspaces.  This will be gengo/v2.  The
+   hand-rolled parsing logic will be removed and replaced with
+   `golang.org/x/tools/go/packages.Load`, which understands workspaces.
+1. Update kube-openapi to use gengo/v2.  TBD as to whether this represents
+   kube-openapi/v2 (that may be simplest).
+1. Introduce workspaces to k/k.
+   1. Add a `go.work` file (which breaks a ton of intra-repo tools).
+   1. Convert non-gengo parts of our build (e.g. `make`, `make test`, many
+      `verify-foo.sh and `update-foo.sh` scripts) to use workspaces (often just
+      removing things like `GO111MODULE=off` or `-mod=mod` decorations).
+   1. Update gengo and kube-openapi versions in k/k (vendored).
+   1. Convert gengo-based tools (in staging/src/k8s.io/code-generator) to
+      gengo/v2 and kube-openapi/v2.
+   1. Fix any remaining tools which had dependencies on the gengo-based tools.
+   1. Remove detritus, including (but not limited to) run-in-gopath.sh,
+      references to GOPATH, workarounds for vendored packages, etc.
+
+Along the way, some tools will need more touch than others.  At least one -
+import-boss - seems best to retool away from gengo entirely.
+
+Somewhere in there, we want to make deeper changes to gengo/v2 to simplify it.
+Some of the framework seemed clean at the time, but nearly a decade later, just
+makes the tools harder to understand.  This could happen after all of this
+work (repeating the sequence) or at the same time, making for larger but more
+complete reworking.
+
+### Go and v2
+
+There's been a lot written about Go's v2 problem.  It can be confusing for
+producers and consumers.  That said, it seems to work for what we want, which
+is to NOT make explicit releases of gengo, even though there's a "v2".
+
+Once consumers switch to v2 (`go get k8s.io/gengo/v2`), they can simply use
+normal Go unversioned updates (`go get -u k8s.io/gengo/v2`).
```

Comment:
Maybe worth summarizing the experiment you've made above wrt v2, and specifically its conclusions with an explicit it works today. I wonder if that's something we can actually rely on in the long term. Maybe worth double checking with go team, if possible?

Reply:
I think we can rely on it. The documentation for it is a bit hard to find, but there is enough of it and now practical evidence (observable behavior is part of an API...) that future Go versions cannot simply change it.

Reply:
I'll add it as an appendix